### ISSUE TITLE
SSL + Haproxy Protocol

### ISSUE THREAD

**Issue by shaggie76** [inspircd/inspircd#1726]
From what I can tell you can't use both SSL and the Haproxy protocol at the same time in InspIRCd.

For our tests with a network partner that required both we had to run a localhost haproxy for SSL termination that forwarded to a localhost-only binding on the IRC server listening for the proxy protocol without SSL; this worked but we found it to be extremely expensive for the CPU -- somehow hosting the SSL in IRC used vastly less CPU for the same number of connections (same ciphers too!).

I checked the source and the way the proxy plugin works it doesn't seem like it can be paired with an SSL binding at all which is a shame.

I'd be happy to discuss our motives for this with more detail in private if possible.

**Comment by SaberUK** [comment]
This is correct. You have to use SSL termination in HAProxy if you want to use it with InspIRCd. It's also recommended that you use UNIX socket listeners with HAProxy as the overhead of using TCP connections is unnecessary.

**Comment by shaggie76** [comment]
Just a few figures of merit why it would be better to merge the two: we cluster 6 IRC processes -- one hub and 5 spokes for clients to connect to. To get better haproxy scaling we spun up 5 processes for termination, mapping each to a spoke. 

Ball-park for 55K connections (11K each spoke), the best CPU load per process we saw was:

14% for SSL-termination in InspIRCd
6% for SSL-termination in haproxy + 12% for raw InspIRCd

So moving termination to haproxy shifts 14% of the CPU load out of InspIRCd and the triples it in another haproxy. 

Note this was with the same ciphers used for both but it was not using unix-domain sockets because it wasn't clear to me how to do that with InspIRCd. This was tested on a real-metal hex-core with hyperthreading.

In general we found the performance to be often radically worse: during periods where users reconnect all at once (after a game update) haproxy CPU load would spike to almost a full core -- dangerously high and something we weren't used to seeing with InspIRCd doing everything. We extrapolated from the data we saw in limited live-fire tests and concluded it would be very risky to use it as an SSL terminator.

**Comment by shaggie76** [comment]
I sent you an email at petpow with some private details -- hope it doesn't get junked.

**Comment by shaggie76** [comment]
Some live-fire data that might interest you:

In our setup we have a hub and spoke configuration (so 5 inspircd spokes behind 5 haproxy processes using Unix Domain Sockets). Here’s what a 10-second averaged top looks like on a bare-metal hex-core Xeon (pid 25023 is the hub):

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
23879 ircd      20   0  380308 313148   8752 S  12.2  0.2  74:28.61 haproxy
23880 ircd      20   0  378728 311508   8752 R  12.2  0.2  75:31.97 haproxy
23881 ircd      20   0  378352 310852   8752 S  12.2  0.2  75:02.69 haproxy
23883 ircd      20   0  381000 313292   8748 S  12.1  0.2  70:48.66 haproxy
23882 ircd      20   0  380848 312804   8748 S  11.9  0.2  73:18.27 haproxy
25125 ircd      20   0 4297356 183900  12520 S  11.4  0.1  68:34.24 inspircd
25091 ircd      20   0 3772620 183128  12516 S  11.2  0.1  69:15.45 inspircd
25159 ircd      20   0 3772768 183120  12368 S  11.2  0.1  68:18.97 inspircd
25193 ircd      20   0 3970408 184744  12532 S  11.1  0.1  68:17.81 inspircd
25057 ircd      20   0 2986488 182640  12628 S  10.9  0.1  68:48.68 inspircd
25023 ircd      20   0  630388 123796  12112 S   4.5  0.1  26:35.82 inspircd
```

Yesterday we had it configured to just do SSL termination in inspircd (same ciphers); with 10% less users we were seeing about 10% less CPU load for the 5 spokes (and 0% haproxy). IE: SSL-offload to haproxy made no significant difference to inspircd and the CPU cost of haproxy is greater than inspircd.

**Comment by shaggie76** [comment]
I've been reading the source to consider our next step here -- my attempts to tune haproxy failed to reduce the overhead to an acceptable level; at the scale we're at it just isn't safe for us to burn CPU on haproxy for SSL termination.

At first glance I don't think it would be too difficult to hack the relevant bits of m_haproxy.cpp into m_ssl_openssl.cpp. I could copy & paste the relevant stuff but it would be a black eye and I doubt you'd accept it in a pull-request.

I could set up m_haproxy.h interfaces to the module and try to share code but again I have my doubts that you'd accept it.

I was also trying to figure out if you could chain these IO hooks -- m_haproxy is a IOHookMiddle but m_ssl_openssl is a IOHook and I'm not sure if I can get them together in the right order.

What I'm most torn about is spending [any more] time on this and having to suffer merging local patches every time we update inspircd. If there's a way forward here that you could recommend I'd appreciate the perspective.

**Comment by shaggie76** [comment]
Related: maybe part of the reason we saw less offload than expected is I think m_haproxy hits the heap for every packet? it calls SendQueue::moveall and ReadData call clear() (ie: free) whenever it's larger than the string reserve (a few dozen bytes I think). Around here we're fans of replacing clear() with resize(0) where appropriate.
